src/createStore.js

```diff
@@ -92,10 +93,20 @@ export default function createStore(reducer, initialState) {
       isPlainObject(action),
       'Actions must be plain objects. Use custom middleware for async actions.'
     );
+    invariant(
+      !isDispatching,
+      'Cannot dispatch in the middle of a dispatch.'
+    );
+
+    isDispatching = true;
 
-    currentState = currentReducer(currentState, action);
-    listeners.forEach(listener => listener());
-    return action;
+    try {
+      currentState = currentReducer(currentState, action);
+      listeners.forEach(listener => listener());
```

Comment:
We want to move this line out. It's OK to call the listeners. The only line we want to guard is `currentState` assignment.